dynawo/sources/Models/CPP/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp

```diff
@@ -0,0 +1,579 @@
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalFType() {
+    fType_[0] = DIFFERENTIAL_EQ;  // differential equations
+    fType_[1] = ALGEBRAIC_EQ;
+    fType_[2] = ALGEBRAIC_EQ;
+    fType_[3] = ALGEBRAIC_EQ;
+    fType_[4] = ALGEBRAIC_EQ;
+    fType_[5] = ALGEBRAIC_EQ;
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::UfRaw() const {
+    return yLocal_[UfRawYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Uf() const {
+    return yLocal_[UfYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::UfRawPrim() const {
+    return ypLocal_[UfRawYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Ur() const {
+    return yLocal_[UrYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Ir() const {
+    return yLocal_[IrYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Ii() const {
+    return yLocal_[IiYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Ui() const {
+    return yLocal_[UiYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::U() const {
+    return sqrt(Ur() * Ur() + Ui() * Ui());
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::U2() const {
+    return Ur() * Ur() + Ui() * Ui();
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::P() const {
+    return yLocal_[PYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Q() const {
+    return yLocal_[QYNum_];
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalF(double /*t*/, propertyF_t type) {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      double UfRawValue = UfRaw();
+      if (UMaxPuReached_ == true) {
+        UfRawValue = UMaxPu_;
+      }
+      if (UMinPuReached_ == true) {
+        UfRawValue = UMinPu_;
+      }
+      if (type == ALGEBRAIC_EQ) {
+        fLocal_[1] = Uf() - UfRawValue;
+        fLocal_[2] = Ir() - (P() * Ur() + Q() * Ui()) / (U() * U());
+        fLocal_[3] = Ii() - (P() * Ui() - Q() * Ur()) / (U() * U());
+        fLocal_[4] = P() - P0_ * pow(U() / Uf(), alpha_);
+        fLocal_[5] = Q() - Q0_ * pow(U() / Uf(), beta_);
+      } else if (type == DIFFERENTIAL_EQ) {
+        fLocal_[0] = Tf_ * UfRawPrim() - (U() - UfRaw());
+      } else {
+        fLocal_[0] = Tf_ * UfRawPrim() - (U() - UfRaw());
+        fLocal_[1] = Uf() - UfRawValue;
+        fLocal_[2] = Ir() - (P() * Ur() + Q() * Ui()) / (U() * U());
+        fLocal_[3] = Ii() - (P() * Ui() - Q() * Ur()) / (U() * U());
+        fLocal_[4] = P() - P0_ * pow(U() / Uf(), alpha_);
+        fLocal_[5] = Q() - Q0_ * pow(U() / Uf(), beta_);
+      }
+    } else {
+      fLocal_[0] = UfRaw();
+      fLocal_[1] = Uf();
+      fLocal_[2] = Ir();
+      fLocal_[3] = Ii();
+      fLocal_[4] = P();
+      fLocal_[5] = Q();
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::setFequations() {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      unsigned int index = 0;
+      fEquationIndex_[index] = std::string("Tf_*UfPrim() - (U() - Uf()):").append(staticId());
```

Comment:
() useless, this is a description of the equation Tf*UfPrim -(u-uf)

Reply:
done